Lib/test/test_free_threading/test_str.py

```diff
@@ -0,0 +1,74 @@
+import sys
+import unittest
+
+from itertools import cycle
+from threading import Event, Thread
+from unittest import TestCase
+
+from test.support import threading_helper
+
+@threading_helper.requires_working_threading()
+class TestStr(TestCase):
+    def test_racing_join_extend(self):
+        '''Test joining a string being extended by another thread'''
+        l = []
+        ITERS = 100
+        READERS = 10
+        done_event = Event()
+        def writer_func():
+            for i in range(ITERS):
+                l.extend(map(str, range(i)))
+                l.clear()
+            done_event.set()
+        def reader_func():
+            while not done_event.is_set():
+                ''.join(l)
+        writer = Thread(target=writer_func)
+        readers = []
+        for x in range(READERS):
+            reader = Thread(target=reader_func)
+            readers.append(reader)
+            reader.start()
+
+        writer.start()
+        writer.join()
+        for reader in readers:
+            reader.join()
+
+    def test_racing_join_replace(self):
+        '''
+        Test joining a string of characters being replaced with ephemeral
+        strings by another thread.
+        '''
+        l = [*'abcdefg']
+        MAX_ORDINAL = 1_000
+        READERS = 20
+
+        def writer_func():
+            for i, c in zip(cycle(range(len(l))),
+                            map(chr, range(128, MAX_ORDINAL))):
+                l[i] = c
+            del l[:]  # Empty list to tell readers to exit
+
+        def reader_func():
+            while True:
+                empty = not l
+                ''.join(l)
+                if empty:
+                    break
```

Comment:
I suggest unrolling the loop a bit so that the threads spend more time calling str.join and less time possibly synchronizing on an event. It will more reliably reproduce the underlying problems.

For example:

```
            while not done_event.is_set():
                ''.join(l)
                ''.join(l)
                ''.join(l)
                ''.join(l)
```

Reply:
Done on both suggestions. On my local box, running in debug mode 100 times, the first test failed on the original code base 100/100 times, the second test failed 91/100 times even with fewer threads involved, so even if future changes to `list` realloc make the first test capable of passing, the second test should catch problems. Good to check-in? I'll start work on applying the macro to other spots where `PySequence_Fast` is used once I'm confident the macro is locked in and I won't have to go back and rename/refactor it.